In [ ]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

In [3]:
import pandas as pd

data = pd.read_csv('/kaggle/input/sms-spam-collection-dataset/spam.csv', encoding='latin-1')

In [4]:
import os

os.listdir('/kaggle/input')

['sms-spam-collection-dataset']

In [5]:
os.listdir('/kaggle/input/sms-spam-collection-dataset')

['spam.csv']

In [6]:
data = data[['v1', 'v2']]
data.columns = ['label', 'message']

In [7]:
data.head()

label                                            message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [8]:
data = data.dropna()

In [9]:
data = data[data['message'].str.strip() != '']

In [10]:
data['label'] = data['label'].map({'ham': 0, 'spam': 1})

In [11]:
data['label'].value_counts()

label
0    4825
1     747
Name: count, dtype: int64

In [13]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer(stop_words='english')
X = vectorizer.fit_transform(data['message'])
y = data['label']

In [14]:
print(X.shape)
print(y.shape)

(5572, 8404)
(5572,)


In [15]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [16]:
print(X_train.shape)
print(X_test.shape)

(4457, 8404)
(1115, 8404)


In [17]:
from sklearn.naive_bayes import MultinomialNB

model = MultinomialNB()
model.fit(X_train, y_train)

MultinomialNB()

In [18]:
y_pred = model.predict(X_test)

In [19]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.9802690582959641


In [20]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred)
print(cm)

[[954  11]
 [ 11 139]]


In [21]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.99      0.99      0.99       965
           1       0.93      0.93      0.93       150

    accuracy                           0.98      1115
   macro avg       0.96      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115



In [24]:
msg = ["Congratulations! You won a free lottery. Call now"]
msg_vector = vectorizer.transform(msg)
prediction = model.predict(msg_vector)

print("Spam" if prediction[0] == 1 else "Ham")

Spam


In [25]:
msg = ["Hey, are we meeting today?"]

## Model Performance

The Naive Bayes classifier achieved an accuracy of **98%** on the test dataset.

- Precision (Spam): 0.93
- Recall (Spam): 0.93
- F1-score (Spam): 0.93

This indicates that the model effectively identifies spam messages while maintaining very low false positives.


## Conclusion

In this project, a machine learning model was developed to classify SMS messages as Spam or Ham.
Using text vectorization and a Naive Bayes classifier, the model achieved high accuracy and reliable performance.

This project demonstrates practical application of:
- Text preprocessing
- Feature extraction
- Machine learning classification
- Model evaluation
